src/core/ext/transport/chttp2/transport/frame.cc

```diff
@@ -0,0 +1,519 @@
+
+ private:
+  SliceBuffer& out_;
+  MutableSlice extra_bytes_;
+};
+
+absl::Status StripPadding(SliceBuffer& payload) {
+  if (payload.Length() < 1) {
+    return absl::InternalError("padding flag set but no padding byte");
+  }
+  uint8_t padding_bytes;
+  payload.MoveFirstNBytesIntoBuffer(1, &padding_bytes);
+  if (payload.Length() < padding_bytes) {
+    return absl::InternalError("padding flag set but not enough padding bytes");
+  }
+  payload.RemoveLastNBytes(padding_bytes);
+  return absl::OkStatus();
+}
+
+absl::StatusOr<Http2DataFrame> ParseDataFrame(const Http2FrameHeader& hdr,
+                                              SliceBuffer& payload) {
+  if (hdr.flags & ~(kFlagEndStream | kFlagPadded)) {
+    return absl::InternalError(
+        absl::StrCat("unsupported data flags: ", hdr.ToString()));
+  }
+
+  if (hdr.stream_id == 0) {
+    return absl::InternalError(
+        absl::StrCat("invalid stream id: ", hdr.ToString()));
+  }
+
+  if (hdr.flags & kFlagPadded) {
+    auto s = StripPadding(payload);
+    if (!s.ok()) return s;
+  }
+
+  return Http2DataFrame{hdr.stream_id, ExtractFlag(hdr.flags, kFlagEndStream),
+                        std::move(payload)};
+}
+
+absl::StatusOr<Http2HeaderFrame> ParseHeaderFrame(const Http2FrameHeader& hdr,
+                                                  SliceBuffer& payload) {
+  if (hdr.flags &
+      ~(kFlagEndHeaders | kFlagEndStream | kFlagPriority | kFlagPadded)) {
+    return absl::InternalError(
+        absl::StrCat("unsupported header flags: ", hdr.ToString()));
+  }
+
+  if (hdr.stream_id == 0) {
+    return absl::InternalError(
+        absl::StrCat("invalid stream id: ", hdr.ToString()));
+  }
+
+  if (hdr.flags & kFlagPadded) {
+    auto s = StripPadding(payload);
+    if (!s.ok()) return s;
+  }
+
+  if (hdr.flags & kFlagPriority) {
+    if (payload.Length() < 5) {
+      return absl::InternalError(
+          absl::StrCat("invalid priority payload: ", hdr.ToString()));
+    }
+    uint8_t trash[5];
+    payload.MoveFirstNBytesIntoBuffer(5, trash);
+  }
+
+  return Http2HeaderFrame{
+      hdr.stream_id, ExtractFlag(hdr.flags, kFlagEndHeaders),
+      ExtractFlag(hdr.flags, kFlagEndStream), std::move(payload)};
+}
+
+absl::StatusOr<Http2ContinuationFrame> ParseContinuationFrame(
+    const Http2FrameHeader& hdr, SliceBuffer& payload) {
+  if (hdr.flags & ~kFlagEndHeaders) {
+    return absl::InternalError(
+        absl::StrCat("unsupported header flags: ", hdr.ToString()));
+  }
+
+  if (hdr.stream_id == 0) {
+    return absl::InternalError(
+        absl::StrCat("invalid stream id: ", hdr.ToString()));
+  }
+
+  return Http2ContinuationFrame{hdr.stream_id,
+                                ExtractFlag(hdr.flags, kFlagEndHeaders),
+                                std::move(payload)};
+}
+
+absl::StatusOr<Http2RstStreamFrame> ParseRstStreamFrame(
+    const Http2FrameHeader& hdr, SliceBuffer& payload) {
+  if (hdr.flags != 0) {
```

Comment:
As per RFC 9113, Unused flags MUST be ignored on receipt

Reply:
fixed